pkg/executor/mem_reader.go

```diff
@@ -1018,46 +1022,78 @@ func (m *memIndexMergeReader) getMemRowsIter(ctx context.Context) (memRowsIter,
 func (m *memIndexMergeReader) getMemRows(ctx context.Context) ([][]types.Datum, error) {
 	r, ctx := tracing.StartRegionEx(ctx, "memIndexMergeReader.getMemRows")
 	defer r.End()
-	tbls := []table.Table{m.table}
-	// [partNum][indexNum][rangeNum]
-	var kvRanges [][][]kv.KeyRange
-	if m.partitionMode {
-		tbls = tbls[:0]
-		for _, p := range m.partitionTables {
-			tbls = append(tbls, p)
+
+	var err error
+	hMap := kv.NewHandleMap()
+	for i, reader := range m.memReaders {
+		// [partNum][rangeNum]
+		var readerKvRanges [][]kv.KeyRange
+		if m.partitionMode {
+			if r, ok := reader.(*memIndexReader); ok && r.index.Global {
+				keyRange, _ := distsql.IndexRangesToKVRanges(m.ctx.GetDistSQLCtx(), r.table.ID, r.index.ID, m.ranges[i])
+				readerKvRanges = [][]kv.KeyRange{keyRange.FirstPartitionRange()}
+			} else {
+				for _, pKeyRanges := range m.partitionKVRanges { // get all keyRanges related to this PartialIndex
+					readerKvRanges = append(readerKvRanges, pKeyRanges[i])
+				}
+			}
+		} else {
+			readerKvRanges = [][]kv.KeyRange{m.indexMergeReader.keyRanges[i]}
+		}
+
+		for j, kr := range readerKvRanges {
```

Comment:
this function looks so long, why remove the original encapsulation here?

Reply:
Because the original loop order is 
```
for each partition {
    for each index {
        for each range {
            ....
        }
    }
    do union or intersection
}
```

Now I changed it to
```
for each index {
    for each partition {
        for each range {
            ....
        }
    }
}

do union or intersection
```

The order of loops is same as `indexMergeExec`. The reason is for global index which is in partition mode, but it only has one partition. And it's hard to create table ranges by `distsql.TableHandlesToKVRanges` function.

Reply:
Done, I divided this function into two functions